File: intro-langchain-python/examples/2-chains-e-processamento/3-runnable-lambda.py
from langchain_core.runnables import RunnableLambda
from langchain.prompts import PromptTemplate
from langchain_openai import ChatOpenAI
from dotenv import load_dotenv
from pydantic import SecretStr
import os
import re

load_dotenv()

def extract_and_analyze_text(text: str) -> dict:
    """
    Função prática que analisa um texto e extrai informações úteis
    para criar um resumo personalizado
    """
    
    # Extrair emails
    email_pattern = r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'
    emails = re.findall(email_pattern, text)
    
    # Contar palavras e sentenças
    words = text.split()
    sentences = text.split('.')
    
    # Identificar palavras-chave importantes
    keywords = ['projeto', 'reunião', 'prazo', 'urgente', 'importante', 'deadline']
    found_keywords = [word for word in keywords if word.lower() in text.lower()]
    
    # Determinar nível de urgência
    urgent_words = ['urgente', 'asap', 'imediatamente', 'crítico', 'emergência']
    urgency_level = sum(1 for word in urgent_words if word.lower() in text.lower())
    
    return {
        "original_text": text,
        "word_count": len(words),
        "sentence_count": len([s for s in sentences if s.strip()]),
        "emails": emails,
        "keywords": found_keywords,
        "urgency_score": urgency_level,
        "text_length": "curto" if len(words) < 20 else "médio" if len(words) < 50 else "longo",
        "has_contact": len(emails) > 0,
        "summary": f"Texto {len(words)} palavras com {len(found_keywords)} palavras-chave importantes"
    }
    
def main():
    print("=== EXEMPLO PRÁTICO: ANÁLISE DE TEXTO COM RUNNABLE LAMBDA COM LANGCHAIN ===\n")
    
    # Criar RunnableLambda para análise de texto
    analyze_runnable = RunnableLambda(extract_and_analyze_text)
    
    # Template que usa os dados analisados
    template = PromptTemplate(
        input_variables=["summary", "keywords", "urgency_score", "text_length"],
        template="""
            Baseado na análise do texto:

            Resumo: {summary}
            Palavras-chave encontradas: {keywords}
            Nível de urgência: {urgency_score}
            Tamanho do texto: {text_length}

            Crie um resumo executivo deste texto e sugira qual a melhor forma de responder ou agir sobre ele.
            """.strip()
    )
    
    # Modelo
    model = ChatOpenAI(
        model="gpt-4o", 
        temperature=0.5,
        #max_completion_tokens=300,
        base_url=os.environ["GITHUB_MODELS_ENDPOINT"],
        api_key=SecretStr(os.environ["GITHUB_MODELS_TOKEN"])
    )
    
    # Chain: análise -> template -> modelo
    analysis_chain = analyze_runnable | template | model
    
    # Texto de exemplo realista
    sample_text = """
        Prezado João,
        
        Preciso urgentemente do relatório do projeto para a reunião de amanhã.
        O prazo é crítico e o cliente estará presente. Por favor, envie para 
        joao.silva@empresa.com até as 14h de hoje.
        
        É muito importante que inclua os dados de performance e o cronograma atualizado.
        
        Obrigada!   
    """
    
    print("Texto para análise:")
    print(sample_text)
    print("\n" + "="*60 + "\n")
    
    # Executar apenas a análise primeiro
    print("Resultado da análise:")
    analysis_response = analyze_runnable.invoke(sample_text)
    
    for key, value in analysis_response.items():
        if key != "original_text": # Não imprimir o texto original
            print(f"{key}: {value}")
    
    print("\n" + "="*60 + "\n")
    
    # Executar a chain completa
    print("Resumo Executivo gerado pelo LLM:")
    try:
        response = analysis_chain.invoke(sample_text)
        print(response.content)
    except Exception as e:
        print(f"Erro ao gerar o resumo executivo: {e}")
    
if __name__ == "__main__":
    main()
    
    

File: intro-langchain-typescript/src/examples/2-chains-e-processamento/3-runnable-lambda.ts
import { ChatOpenAI } from "@langchain/openai";
import { PromptTemplate } from "@langchain/core/prompts";
import { RunnableLambda } from "@langchain/core/runnables";
import {
  loadEnvironment,
  getApiKey,
  getEndpoint,
  displayResult,
} from "../../utils/helpers";

loadEnvironment();

// Interface para o resultado da análise
interface TextAnalysis {
  original_text: string;
  word_count: number;
  sentence_count: number;
  emails: string[];
  keywords: string[];
  urgency_score: number;
  text_length: "curto" | "médio" | "longo";
  has_contact: boolean;
  summary: string;
}

function extractAndAnalyzeText(text: string): TextAnalysis {
  // Extrair emails usando regex
  const emailPattern = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g;
  const emails = text.match(emailPattern) || [];

  // Contar palavras e sentenças
  const words = text.split(/\s+/).filter((word) => word.length > 0);
  const sentences = text
    .split(".")
    .filter((sentence) => sentence.trim().length > 0);

  // Identificar palavras-chave importantes
  const keywords = [
    "projeto",
    "reunião",
    "prazo",
    "urgente",
    "importante",
    "deadline",
  ];
  const foundKeywords = keywords.filter((keyword) =>
    text.toLowerCase().includes(keyword.toLowerCase())
  );

  // Determinar nível de urgência
  const urgentWords = [
    "urgente",
    "asap",
    "imediatamente",
    "crítico",
    "emergência",
  ];
  const urgencyLevel = urgentWords.reduce(
    (count, word) =>
      count + (text.toLowerCase().includes(word.toLowerCase()) ? 1 : 0),
    0
  );

  // Determinar tamanho do texto
  let textLength: "curto" | "médio" | "longo";
  if (words.length < 20) {
    textLength = "curto";
  } else if (words.length <= 50) {
    textLength = "médio";
  } else {
    textLength = "longo";
  }

  return {
    original_text: text,
    word_count: words.length,
    sentence_count: sentences.length,
    emails: emails,
    keywords: foundKeywords,
    urgency_score: urgencyLevel,
    text_length: textLength,
    has_contact: emails.length > 0,
    summary: `Texto ${words.length} palavras com ${foundKeywords.length} palavras-chave importantes`,
  };
}

async function main(): Promise<void> {
  try {
    console.log(
      "=== EXEMPLO PRÁTICO: ANÁLISE DE TEXTO COM RUNNABLE LAMBDA (LangChain.js) ===\n"
    );

    // Criar RunnableLambda para análise de texto
    const analyzeRunnable = new RunnableLambda({
      func: extractAndAnalyzeText,
    });

    // Template que usa os dados analisados
    const template = new PromptTemplate({
      inputVariables: ["summary", "keywords", "urgency_score", "text_length"],
      template: `
        Baseado na análise do texto:

        Resumo: {summary}
        Palavras-chave encontradas: {keywords}
        Nível de urgência: {urgency_score}
        Tamanho do texto: {text_length}

        Crie um resumo executivo deste texto e sugira qual a melhor forma de responder ou agir sobre ele
      `.trim(),
    });

    const model = new ChatOpenAI({
      model: "gpt-4o",
      temperature: 0.5,
      maxTokens: 500,
      configuration: {
        baseURL: getEndpoint("GITHUB_MODELS_ENDPOINT"),
        apiKey: getApiKey("GITHUB_MODELS_TOKEN"),
      },
    });

    // Chain: análise -> template -> modelo
    const analysisChain = analyzeRunnable.pipe(template).pipe(model);

    // Texto de exemplo realista
    const sampleText = `
      Prezado João,
      
      Preciso urgentemente do relatório do projeto para a reunião de amanhã.
      O prazo é crítico e o cliente estará presente. Por favor, envie para 
      joao.silva@empresa.com até as 14h de hoje.
      
      É muito importante que inclua os dados de performance e o cronograma atualizado.
      
      Obrigada!     
    `;

    console.log("Texto para Análise");
    console.log(sampleText);
    console.log("\n" + "=".repeat(60) + "\n");

    // Executar apenas a análise primeiro
    console.log("Resultado da análise do texto");
    const analysisResponse = await analyzeRunnable.invoke(sampleText);

    // Exibir resultados da análise (exceto o texto original)
    Object.entries(analysisResponse).forEach(([key, value]) => {
      if (key !== "original_text") {
        console.log(
          `  ${key}: ${Array.isArray(value) ? value.join(", ") : value}`
        );
      }
    });

    console.log("\n" + "=".repeat(60) + "\n");

    // Executar a chain completa
    console.log("Resumo Executivo gerado pelo LLM");
    try {
      const response = await analysisChain.invoke(sampleText);
      displayResult("Análise do Texto com RunnableLambda", response.content);
    } catch (error) {
      console.error(
        "Erro ao gerar o resumo executivo:",
        error instanceof Error ? error.message : error
      );
    }
  } catch (error) {}
}

main();
